===================================================================
RCS file: /cvs/gnome/gtk+/gdk-pixbuf/io-bmp.c,v
retrieving revision 1.46
retrieving revision 1.46.2.2
diff -u -r1.46 -r1.46.2.2
--- io-bmp.c	2005/01/04 15:47:02	1.46
+++ io-bmp.c	2005/03/28 04:12:32	1.46.2.2
@@ -219,7 +219,19 @@
 static gboolean grow_buffer (struct bmp_progressive_state *State,
                              GError **error)
 {
-  guchar *tmp = g_try_realloc (State->buff, State->BufferSize);
+  guchar *tmp;
+
+  if (State->BufferSize == 0) {
+    g_set_error (error,
+		 GDK_PIXBUF_ERROR,
+		 GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
+		 _("BMP image has bogus header data"));
+    State->read_state = READ_STATE_ERROR;
+    return FALSE;
+  }
+
+  tmp = g_try_realloc (State->buff, State->BufferSize);
+
   if (!tmp) {
     g_set_error (error,
 		 GDK_PIXBUF_ERROR,
@@ -228,6 +240,7 @@
     State->read_state = READ_STATE_ERROR;
     return FALSE;
   }
+
   State->buff = tmp;
   return TRUE;
 }
@@ -1031,7 +1044,7 @@
 			gint new_y = MIN (context->compr.y, context->Header.height);
 			(*context->updated_func) (context->pixbuf,
 						  0,
-						  y,
+						  context->Header.height - new_y,
 						  context->Header.width,
 						  new_y - y,
 						  context->user_data);
